bindings/node/test/clientEncryption.test.js

```diff
@@ -815,4 +829,115 @@ describe('ClientEncryption', function () {
   it('should provide the libmongocrypt version', function () {
     expect(ClientEncryption.libmongocryptVersion).to.be.a('string');
   });
+
+  describe('createEncryptedCollection()', () => {
+    /** @type {InstanceType<ClientEncryption>} */
+    const client = new MockClient();
+    let clientEncryption;
+    let db;
+    const collectionName = 'secure';
+
+    beforeEach(async function () {
+      clientEncryption = new ClientEncryption(client, {
+        keyVaultNamespace: 'client.encryption',
+        kmsProviders: { local: { key: Buffer.alloc(96, 0) } }
+      });
+
+      db = client.db('createEncryptedCollectionDb');
+      await db.dropDatabase().catch(() => null);
+    });
+
+    afterEach(async () => {
+      sinon.restore();
+      await db.dropDatabase().catch(() => null);
+    });
+
+    it('throws if options are omitted', async () => {
+      const error = await clientEncryption
+        .createEncryptedCollection(db, collectionName)
+        .catch(error => error);
+      expect(error).to.be.instanceOf(TypeError, /Cannot destructure/);
+    });
+
+    it('throws if options.encryptedFields are omitted', async () => {
+      const error = await clientEncryption
+        .createEncryptedCollection(db, collectionName, {})
+        .catch(error => error);
+      expect(error).to.be.instanceOf(MongoCryptError);
+    });
+
+    it('throws MongoCryptCreateEncryptedCollectionError if createDataKey rejects', async () => {
```

Comment:
The logic in the method is incorrect.  From the spec:
```
Let EF' be a copy of EF. Update EF' in the following manner:
Let Fields be the "fields" element within EF'.
If Fields is present and is an array value, then for each element F of Fields:
If F is not a document element, skip it.
Otherwise, if F has a "keyId" named element K and K is a null value:
Let D be the result of CE.createDataKey(kmsProvider, dkOpts).
If generating D resulted in an error E, the entire CreateEncryptedCollection must now fail with error E. Return the partially-formed EF' with the error so that the caller may know what datakeys have already been created by the helper.
```

So first we deep copy the whole EF object, then we update elements as we encounter them.  Consequently, if an error is encountered, we return the whole copied EF document (which includes any fields that have not yet been processed).  Your logic pushes new fields into the array, so when an error is encountered, only processed keys are returned in the returned `encryptedFieldsMap`.

So test cases we need to ensure we test here (I would have specific tests for each test case, perhaps all wrapped in a context block)

when there is an error thrown

- the root cause of the error is assigned to the mongo crypt error
- the encrypted fields map is present on the error
- the encrypted fields map is the same length as the original and contains documents in the same order
- the encrypted fields map only has data keys filled out for successful data key creations